### ISSUE TITLE
update to notify 5.0

### ISSUE THREAD

**Issue by cmyr** [xi-editor/xi-editor#1158]
A vague tracking issue: a major update to the notify crate is in the works, and we should investigate what will be involved in migrating to use it. I haven't looked into this at all, but there are [some docs here](https://github.com/passcod/notify/tree/v5-readme#notify).

**Comment by sjoshid** [comment]
I can pick this up as Im kinda familiar with what's coming.

**Comment by sjoshid** [comment]
@cmyr 
I made changes (need to make changes to tests) according to the new event system coming up in notify 5.0 but I have one question. 5.0 is using crossbeam channels and Im wondering if we can use the new crate to do things differently in xi. This is what I got from the author of notify:

> notify now uses crossbeam channels, which means you can start
multiple watchers, configured differently, on the *same* reporting channel.

If I understand this correctly, we dont have to use our custom wrapper in `watcher.rs` which I believe was written to support same use case: multiple watchers watching same path. 

But this begs the question: is it worth it? We would be just refactoring a bunch to things to do what we do now.

**Comment by cmyr** [comment]
@sjoshid Excellent question, and I don't have any strong opinions. If things are working, currently, I don't think it makes sense to do a big rewrite (unless you're just curious, in which case go ahead!) but if there are any clear advantages to another pattern I'm open to that too?

**Comment by sjoshid** [comment]
@cmyr I dont see any obvious advantage in doing the refactoring but I'll look at it one more time. 
On a different topic, the new event system in 5.0 can give us more granular event types for eg., CREATE event not just tells us something was created but also tells use WHAT was created: FILE or FOLDER. 
Where xi stands currently I dont know if we should care about FOLDER events. But I do see potential for future where xi might allow us to open folders as well. And when it does we need to watch for CREATE_FOLDER also. 
Do you want me to incorporate folder creation events as part of this issue?

**Comment by cmyr** [comment]
We would care about folder events if, for instance, a new plugin package were added. I think this works okay currently, but I'm not sure?

**Comment by sjoshid** [comment]
Sounds good.

**Comment by sjoshid** [comment]
Just an update on my progress....I got a good sense of what the new event classification system does in 5.0. Main issue Im tracking is https://github.com/passcod/notify/issues/187. There will be more changes coming in as part of 187. 
Putting this one on hold till then.

**Comment by passcod** [comment]
5.0.0-pre.0 is out, which may or may not be what you want... more new event goodies are coming later, but I decided to scrape the entire debouncing and rebuild it instead of slogging through and producing a very dodgy solution cobbled from a bit of the old and a bit of the new, so pre.0 is everything that's reasonable solid before that.

**Comment by sjoshid** [comment]
@passcod thanks for the update. I got a chance to read the release notes but it says 
> be aware that the API will change dramatically between prereleases
Last I checked Event API already had some big changes and if it's going to change dramatically between prereleases...shouldn't we just wait for final release or 5.0.0.pre.5 or something? Im just trying to avoid refactoring again next month.

**Comment by passcod** [comment]
The idea is I'm not likely to be that quick. If you peg to `=5.0.0-pre.0` and ignore all the subsequent prereleases you won't need to refactor until final release... which may be nearer the end of the year, and the goal is to rebuild up to the same debounced API. The internals will change, and the immediate API will change (_has_ changed, if you look at `main`), but the debounced API was fine and will _not_ change significantly.

**Comment by sjoshid** [comment]
Sounds good.
@cmyr I can resume work on this asap.

**Comment by Cogitri** [comment]
Maybe you can make a draft PR for this to make it easier to track progress on this?

**Comment by sjoshid** [comment]
Right. I did some commits to my fork but realized that notify was going to change. Now that it's stable I can do a draft PR.

**Comment by cmyr** [comment]
sounds good!

**Comment by sjoshid** [comment]
I made some changes and ran all tests which are successful. I have questions on changing preference file. 

- What is the expected behavior if I _rename_ prefrences.xiconfig? I think it is invalid testcase as it is Xi's **THE** pref file. Also it is hardcoded. But I maybe wrong. 
-  What Im not sure about are the comments on top of pref file. 
> These settings can also be overridden for a particular syntax;  
add your overrides in (for example) '$XI_CONFIG/rust.xiconfig'. 

So I created a file Swift.xiconfig in same location as the pref file and overrode `word_wrap = true`. When I try to save that file I get errors. See image. I tried with rust.xiconfig as well. Same error. 

![image](https://user-images.githubusercontent.com/6751090/60401900-0717be00-9b56-11e9-8891-a9d4d7eb934c.png)

- Digging deeper I found that the file name has to `<languageId>.xiconfig`. Where can i find list of language ids?
Important note: I get those errors with XI's master as well. Is it a known issue?

**Comment by Cogitri** [comment]
> What is the expected behavior if I rename prefrences.xiconfig? I think it is invalid testcase as it is Xi's THE pref file

Yup, I don't think we should pay attention to that, since it's hard to implement and potentially confusing (e.g. if the user renames it to preferences.xiconfig.orig and creates a new preferences.xiconfig, what file should we track?).

>  I get errors. See image. 

I don't have a mac to validate this, but at least with `Tau` on Linux I don't get such errors. I guess @cmyr  might know more.

> Digging deeper I found that the file name has to <languageId>.xiconfig. Where can i find list of language ids?

https://github.com/xi-editor/xi-editor/blob/3169c041b8befff3384dc1cc8878128a9a837240/rust/syntect-plugin/manifest.toml

In _theory_ other plugins could also provide other langs, but I think as of right now syntect covers about everything.
